hd/hd.1
.\"
.\" Sccsid @(#)hd.1	1.8 (gritter) 12/5/04
.TH HD 1XNX "12/5/04" "Heirloom Toolchest" "XENIX System Compatibility"
.SH NAME
hd \- (XENIX) display files in hexadecimal format
.SH SYNOPSIS
\fBhd\fR
[\fB\-acbwlAxdov\fR]
[\fB\-t\fR]
[\fB\-s\fI\ offset\fR]
[\fB\-n\fI\ count\fR]
[\fIfiles\fR]
.SH DESCRIPTION
The
.I hd
command
displays the contents
of the given
.I files
(or of standard input
if no such argument is present)
in hexadecimal, decimal or octal format,
16 bytes per line.
It optionally displays printable ASCII characters in the rightmost column.
The file address is printed in the leftmost column.
.PP
Output format selection is done
with options; each format selection option consists
(optionally) of a format selector and a base selector.
.PP
The format selectors are as follows;
if no format selector is present,
the base selector is applied to all available format selectors:
.TP
.B \-b
Separates the input in bytes;
each byte is printed as a numerical value.
.TP
.B \-c
Separates the input in bytes;
each byte that is also a printable character
is printed as such,
certain characters are printed as C language escapes,
others as a numerical value.
.TP
.B \-w
Separates the input in 16-bit words;
each word is printed as a numerical value.
.TP
.B \-l
Separates the input in 32-bit long words;
each word is printed as a numerical value.
.PP
The following base selectors are accepted;
if the base selector is omitted,
all base selectors are applied to the chosen format selectors:
.TP
.B \-o
Selects octal output.
.TP
.B \-d
Selects unsigned decimal output.
.TP
.B \-x
Selects hexadecimal output.
.PP
The format of the addresses at the left
can also be selected by using the
.TP
.B \-a
option followed by one of the base selectors;
only one base selector is applied.
The default address base is hexadecimal.
.PP
The following options also affect the output format:
.TP
.B \-A
Causes bytes that are printable characters to be displayed
as such at the right; nonprintable bytes are printed as `.'.
.TP
.B \-t
Text file format; overrides all other format options except
.IR \-a .
Each line of the input is preceded by its address;
printable characters are displayed as such,
except for `\e', `^', and '~', which are prefixed by a `\e' character.
ASCII control characters are prefixed by `^'.
Bytes with the highest bit set that do not form a printable character
are indicated by '~',
followed by the corresponding ASCII character as above.
.TP
.B \-v
Unless
.I \-t
is used,
a set of immediately following indentical 16-byte sets
is abbreviated by printing `*' for the second and further ones.
This option inhibits this behavior
and causes all lines to be displayed.
This option is an extension.
.PP
If no output selection is present, the default is
.IR \-bxA .
.PP
The following options affect the handling of input files:
.TP
\fB\-s\fI\ offset\fR
Selects a starting offset within each file.
.I Offset
is interpreted as a decimal value
unless prefixed by
.BR 0 ,
which causes interpretation as an octal value,
or
.BR 0x ,
which causes interpretation as a hexadecimal value.
The suffixes
.BR w ,
.BR l ,
.BR b ,
and
.B k
cause multiplication by
2 (`words'),
4 (`long words'),
512 (`blocks'),
or
1024 (`kilobytes'),
respectively.
Value and suffix may be separated by a `*' character
to suppress interpretation of `b' as a hexadecimal digit.
.TP
\fB\-n\fI\ count\fR
Causes only
.I count
characters of the file to be displayed;
.I count
is handled as the
.I offset
for
.I \-s
described above.
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_CTYPE
Selects the set of printable single-byte characters
for the
.I \-A
and
.I \-c
options
and the set of printable characters for the
.I \-t
option.
.SH "SEE ALSO"
od(1)


CHANGES
Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.
* cpio: Added a "-D" option to make it usable in combination with SVR4
  packaging tools. This option is not intended for general use, and is
  thus not documented.
* The package build target "make pkg" for Solaris and Open UNIX is no
  longer supported.

Release 070115
* deroff: Now recognizes the ".do" and ".xflag" requests as well as an "-xN"
  command line option to handle Heirloom nroff/troff extensions.
* ed: Substituting only a specific occurence of the pattern, as in the
  command "s/x/y/3", now works correctly if the substitution "y" is not
  equal in length to the pattern match "x" (Derek Thomas).
* ps: Use the Linux 2.6 /proc/status "SleepAVG" field to compute the value
  for the "C" column (cpu usage for scheduling).
* priocntl: The "FF", "B", and "ISO" class names are now accepted with
  the "-e" option too.
* file: UTF-8 characters in troff input files are now recognized.
* cp: The "-s" option now displays correct sizes and related statistics
  for special files with st_size == 0.
* grep: With the "-z" option, decompressor children processes are now
  properly waited for, avoiding failures when processing many files.
* su: Works with PAM on Mac OS X now (Stefan Tramm).
* ps: Handles dead processes as on other platforms now (Stefan Tramm).
* pgrep: Ignores dead processes now (Stefan Tramm).
* spell: A value of the wrong type that was passed to scanf() could lead
  to build errors on 64-bit platforms (Jeffrey H. Johnson).
* ps, pgrep, whodo: Works on FreeBSD 6 now.
* df, find, ps: Changes have been made to build on NetBSD 3 (Aleksey Cheusov).
* Changes have been made to build on OpenBSD 3.8 (Aleksey Cheusov).
* Changes have been made to build with GNU make 3.81.

Release 060122
* The tools have been ported to Mac OS X 10.4.3 (patches by Stefan Tramm).
* The tools have been ported to DragonFlyBSD 1.3.7-DEVELOPMENT.

Release 051118
* rm: The output of the question before descending to a directory with
  the "-i" option has been fixed.
* /usr/ucb/ps: The "-a" option has been fixed; it incorrectly printed a
  list of all processes since release 050411.
* getconf: /usr/ucb is now put at the end of the Heirloom directory list
  for the PATH variable.
* The "-man" macro no longer produces cut marks when used with troff.
* tr: When evaluating '[:class:]' expressions, the last position of the
  coded character set was not included; this has been fixed.
* ps: The "-R" option (resource usage format), which had been present
  for some time, has been documented.
* RPM package builds now require the new "Heirloom Development Tools"
  package from <http://heirloom.sourceforge.net/devtools.html> to be
  installed.

Release 050815
* The "getconf" utility has been added (derived from OpenSolaris code).
* The "mesg" and "pathchk" utilities have been added.
* tr: A new implementation that is based on 6th and 7th edition code is
  now used. It stores the translation tables in arrays of variable size.
  This results in better performance, under some circumstances at the
  expense of higher memory consumption.
* printf: Handle unsigned values larger than INT_MAX correctly with "%u",
  "%o", "%x", and "%X".
* printf: Handle SVID3 "position" fields in format specifications, such
  as "%7$s", that specify which argument is used next.
* pg: If a line does not fit on the screen, do not print a newline to
  force wraparound unless it is really necessary. This allows to select
  words across the wrap for copy-and-paste in an xterm.
* pg: Discard the ESC[48m sequence on a terminal that supports colors.
* ed: Do not print "expecting `w'" with the "q" command after a zero-byte
  file has been written.
* man: The tmac.an macros no longer use "@" as an escape character; its
  previous use lead to weird output with Unix nroff when an "@" appeared
  in titles or hanging labels.
* man: The "TCAT" post processor is now invoked separately for each manual
  entry. Previously, when the -t option was given with multiple entries and
  TCAT=dpost, only the first entry was actually printed.
* Some issues with building on AIX have been fixed.

Release 050713
* The 'echo' and 'test' commands have been added as standalone versions of
  those of the Heirloom Bourne shell.
* The 'false', 'kill', 'printf', 'pwd', and 'true' commands have been added.
* ed: Copying a larger number of lines with the 't' command could cause a
  segmentation fault because some pointers were not updated after calling
  realloc().
* ed: Do not warn about an 'input error' anymore when reading a file that
  lacks the terminating newline.
* ed: Disallow the 'u' command if no change has been made yet after a file
  specified on the command line has been read.

hd/hd.c
/*
 * hd - display files in hexadecimal format
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, September 2003.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)hd.sl	1.12 (gritter) 5/29/05";

#include <sys/types.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <libgen.h>
#include <stdarg.h>
#include <locale.h>
#include <ctype.h>
#include <wctype.h>
#include <wchar.h>
#include <inttypes.h>
#include <limits.h>
#include "atoll.h"
#include "mbtowi.h"

#ifdef	__GLIBC__
#ifdef	_IO_getc_unlocked
#undef	getc
#define	getc(f)		_IO_getc_unlocked(f)
#endif
#ifdef	_IO_putc_unlocked
#undef	putchar
#define	putchar(c)	_IO_putc_unlocked(c, stdout)
#endif
#endif	/* __GLIBC__ */

enum	base {
	BASE_0 = 00,
	BASE_X = 01,
	BASE_D = 02,
	BASE_O = 04
};

union	block {
	int8_t	b_c[16];
	int16_t	b_w[8];
	int32_t	b_l[4];
};

static const struct fm {
	int	f_fmt;
	enum base	f_base;
	char	f_width;
	char	f_align[3];
	const char	*f_prf;
} ofmt[] = {
	{ 'b', BASE_X,  2, {2,5,11}, "%02x" },
	{ 'b', BASE_D,  3, {3,7,15}, "%3u" },
	{ 'b', BASE_O,  3, {3,7,15}, "%03o" },
	{ 'c', BASE_X,  2, {2,5,11}, "%02x" },
	{ 'c', BASE_D,  3, {3,7,15}, "%3u" },
	{ 'c', BASE_O,  3, {3,7,15}, "%03o" },
	{ 'w', BASE_X,  4, {0,4, 9}, "%04x" },
	{ 'w', BASE_D,  5, {0,5,11}, "%5u" },
	{ 'w', BASE_O,  6, {0,6,13}, "%06o" },
	{ 'l', BASE_X,  8, {0,0, 8}, "%08lx" },
	{ 'l', BASE_D, 10, {0,0,10}, "%10lu" },
	{ 'l', BASE_O, 11, {0,0,11}, "%011lo" },
	{   0, BASE_0,  0, {0,0, 0}, NULL }
};

static int		Aflag;	/* print ASCII at right */
static enum base	aflag;	/* address format specifier */
static enum base	bflag;	/* byte format specifier */
static enum base	cflag;	/* print ASCII at center */
static enum base	lflag;	/* long (32 bit) format specifier */
static long long	nflag;	/* number of bytes to process */
static long long	sflag;	/* start offset */
static int		tflag;	/* print text file */
static int		vflag;	/* no '*' for identical lines */
static enum base	wflag;	/* word (16 bit) format specifier */
static char		align[3];
static const char	*progname;
static int		status;
static int		mb_cur_max;

static void		usage(void);
static void		flag(int);
static void		base(enum base, enum base *);
static long long	count(const char *);
static void		usage(void);
static void		diag(const char *, ...);
static void		hd(FILE *);
static void		prna(long long);
static void		prnb(union block *, int);
static void		line(union block *, int, int, enum base, int);
static const struct fm	*getfmt(int, enum base);
static void		getalign(void);
static void		prnt(FILE *, long long);
static void		prnc(int);
static char		*wcget(FILE *fp, wint_t *wc, int *len);

int
main(int argc, char **argv)
{
	FILE	*fp;
	int	i, j;

	progname = basename(argv[0]);
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	for (i = 1; i < argc && argv[i][0] == '-'; i++) {
		switch (argv[i][1]) {
		case 's':
			if (argv[i][2])
				sflag = count(&argv[i][2]);
			else if (++i < argc)
				sflag = count(argv[i]);
			else
				usage();
			break;
		case 'n':
			if (argv[i][2])
				nflag = count(&argv[i][2]);
			else if (++i < argc)
				nflag = count(argv[i]);
			else
				usage();
			break;
		default:
			for (j = 1; argv[i][j]; j++)
				flag(argv[i][j]&0377);
			flag(0);
		}
	}
	if (tflag && (Aflag|bflag|cflag|lflag|wflag))
		diag("-t flag overrides other flags");
	if ((Aflag|bflag|cflag|lflag|wflag) == 0)
		Aflag = 1;
	if ((bflag|cflag|lflag|wflag) == 0)
		bflag = BASE_X;
	getalign();
	if (i < argc) {
		j = i+1 < argc;
		do {
			if (access(argv[i], R_OK) < 0) {
				diag("cannot access %s", argv[i]);
				continue;
			}
			if ((fp = fopen(argv[i], "r")) == NULL) {
				diag("open of %s failed", argv[i]);
				continue;
			}
			if (j)
				printf("%s:\n", argv[i]);
			hd(fp);
			fclose(fp);
			if (i+1 < argc)
				printf("\n");
		} while (++i < argc);
	} else
		hd(stdin);
	return status;
}

static void
flag(int c)
{
	static enum base	*basep;

	switch (c) {
	case '\0':
		if (basep && basep != &aflag && *basep == BASE_0)
			*basep |= BASE_O|BASE_D|BASE_X;
		basep = NULL;
		break;
	case 'a':
		basep = &aflag;
		break;
	case 'b':
		basep = &bflag;
		break;
	case 'w':
		basep = &wflag;
		break;
	case 'l':
		basep = &lflag;
		break;
	case 'c':
		basep = &cflag;
		break;
	case 'A':
		Aflag = 1;
		break;
	case 'x':
		base(BASE_X, basep);
		break;
	case 'd':
		base(BASE_D, basep);
		break;
	case 'o':
		base(BASE_O, basep);
		break;
	case 't':
		tflag = 1;
		break;
	case 'v':
		vflag = 1;
		break;
	default:
		usage();
	}
}

static void
base(enum base b, enum base *basep)
{
	if (basep) {
		if (basep == &aflag)
			*basep = b;
		else
			*basep |= b;
	} else {
		if (aflag == BASE_0)
			aflag |= b;
		cflag |= b;
		bflag |= b;
		wflag |= b;
		lflag |= b;
	}
}

static long long
count(const char *s)
{
	long long	c;
	int	bs = 10;
	char	*x;

	if (s[0] == '0' && s[1] == 'x') {
		bs = 16;
		s += 2;
	} else if (s[0] == '0') {
		bs = 8;
		s++;
	}
	c = strtoll(s, &x, bs);
	s = x;
	if (*s == '*')
		s++;
	switch (*s) {
	case 'w':
		c *= 2;
		s++;
		break;
	case 'l':
		c *= 4;
		s++;
		break;
	case 'b':
		c *= 512;
		s++;
		break;
	case 'k':
		c *= 1024;
		s++;
		break;
	}
	if (*s) {
		diag("bad count/offset value");
		exit(3);
	}
	return c;
}

static void
usage(void)
{
	fprintf(stderr, "usage: %s [-acbwlAxdo] [-t] [-s offset[*][wlbk]] "
			"[-n count[*][wlbk]] [file] ...\n",
			progname);
	exit(2);
}

static void
diag(const char *fmt, ...)
{
	va_list	ap;

	fprintf(stderr, "%s: ", progname);
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	fprintf(stderr, "\n");
	status |= 1;
}

static void
hd(FILE *fp)
{
	long long	of = 0, rd = 0;
	union block	b, ob;
	size_t	n, m, on = 0;
	int	star = 0;

	if (sflag)
		while (of < sflag) {
			getc(fp);
			of++;
		}
	if (tflag) {
		prnt(fp, of);
		return;
	}
	do {
		if (nflag == 0 || rd + sizeof b.b_c < nflag)
			m = sizeof b.b_c;
		else
			m = nflag - rd;
		if ((n = fread(b.b_c, 1, m, fp)) > 0) {
			if (!vflag && n==on && memcmp(b.b_c, ob.b_c, n) == 0) {
				if (star == 0)
					printf("*\n");
				star = 1;
			} else {
				star = 0;
				prna(of);
				if (n < sizeof b.b_c)
					memset(&b.b_c[n], 0, sizeof b.b_c - n);
				prnb(&b, n);
			}
		}
		rd += n;
		of += n;
		on = n;
		ob = b;
	} while (n == m && (nflag == 0 || rd < nflag));
	prna(of);
	putchar('\n');
}

static void
prna(long long n)
{
	switch (aflag) {
	case BASE_O:
		printf("%06llo", n);
		break;
	case BASE_D:
		printf("%05llu", n);
		break;
	case BASE_0:
	case BASE_X:
		printf("%04llx", n);
		break;
	}
}

static void
prnb(union block *bp, int n)
{
	int	cnt = 0;

	if (cflag&BASE_X)
		line(bp, n, 'c', BASE_X, cnt++);
	if (cflag&BASE_D)
		line(bp, n, 'c', BASE_D, cnt++);
	if (cflag&BASE_O)
		line(bp, n, 'c', BASE_O, cnt++);
	if (bflag&BASE_X)
		line(bp, n, 'b', BASE_X, cnt++);
	if (bflag&BASE_D)
		line(bp, n, 'b', BASE_D, cnt++);
	if (bflag&BASE_O)
		line(bp, n, 'b', BASE_O, cnt++);
	if (wflag&BASE_X)
		line(bp, n, 'w', BASE_X, cnt++);
	if (wflag&BASE_D)
		line(bp, n, 'w', BASE_D, cnt++);
	if (wflag&BASE_O)
		line(bp, n, 'w', BASE_O, cnt++);
	if (lflag&BASE_X)
		line(bp, n, 'l', BASE_X, cnt++);
	if (lflag&BASE_D)
		line(bp, n, 'l', BASE_D, cnt++);
	if (lflag&BASE_O)
		line(bp, n, 'l', BASE_O, cnt++);
}

static void
line(union block *bp, int n, int fmt, enum base base, int cnt)
{
	int	c, i, j, k, col = 0;
	const char	*cp;
	const struct fm	*fmp;

	putchar('\t');
	i = 0;
	switch (fmt) {
	case 'l':
		fmp = getfmt('l', base);
		for (j = i/4; j < (n>>2); j++, i += 4) {
			if (col > 0) {
				putchar(' ');
				col++;
			}
			if (i == 8) {
				putchar(' ');
				col++;
			}
			for (k = fmp->f_width; k < align[2]; k++) {
				putchar(' ');
				col++;
			}
			col += printf(fmp->f_prf,(long)(bp->b_l[j]&0xffffffff));
		}
		if (i == n)
			break;
		/*FALLTHRU*/
	case 'w':
		fmp = getfmt('w', base);
		for (j = i/2; j < (n>>1); j++, i += 2) {
			if (col > 0) {
				putchar(' ');
				col++;
			}
			if (i == 8) {
				putchar(' ');
				col++;
			}
			for (k = fmp->f_width; k < align[1]; k++) {
				putchar(' ');
				col++;
			}
			col += printf(fmp->f_prf, (int)(bp->b_w[j]&0177777));
		}
		if (i == n)
			break;
		/*FALLTHRU*/
	case 'b':
		fmp = getfmt('b', base);
		for (j = i; j < n; j++, i++) {
			if (col > 0) {
				putchar(' ');
				col++;
			}
			if (i == 8) {
				putchar(' ');
				col++;
			}
			for (k = fmp->f_width; k < align[0]; k++) {
				putchar(' ');
				col++;
			}
			col += printf(fmp->f_prf, bp->b_c[j]&0377);
		}
		break;
	case 'c':
		fmp = getfmt('c', base);
		for (i = 0; i < n; i++) {
			if (col > 0) {
				putchar(' ');
				col++;
			}
			if (i == 8) {
				putchar(' ');
				col++;
			}
			for (k = fmp->f_width; k < align[0]; k++) {
				putchar(' ');
				col++;
			}
			c = bp->b_c[i]&0377;
			cp = NULL;
			if (c == '\b')
				cp = "\\b";
			else if (c == '\t')
				cp = "\\t";
			else if (c == '\n')
				cp = "\\n";
			else if (c == '\f')
				cp = "\\f";
			else if (c == '\r')
				cp = "\\r";
			else if (!isprint(c)) {
				col += printf(fmp->f_prf, c);
			} else {
				if (base != BASE_X) {
					putchar(' ');
					col++;
				}
				col += printf(" %c", c);
			}
			if (cp) {
				if (base != BASE_X) {
					putchar(' ');
					col++;
				}
				printf(cp);
			}
		}
		break;
	}
	if (cnt == 0 && Aflag) {
		while (col++ < 51)
			putchar(' ');
		for (i = 0; i < n; i++) {
			if ((bp->b_c[i]&0340) == 0 || bp->b_c[i] == 0177 ||
					!isprint(bp->b_c[i]&0377))
				putchar('.');
			else
				putchar(bp->b_c[i]&0377);
		}
	}
	putchar('\n');
}

static const struct fm *
getfmt(int fmt, enum base base)
{
	int	i;

	for (i = 0; ofmt[i].f_fmt; i++)
		if (ofmt[i].f_fmt == fmt && ofmt[i].f_base == base)
			return &ofmt[i];
	return NULL;
}

static void
getalign(void)
{
	int	i, j;
	enum base	*basep;

	for (i = 0; ofmt[i].f_fmt; i++) {
		switch (ofmt[i].f_fmt) {
		case 'b':
			basep = &bflag;
			break;
		case 'c':
			basep = &cflag;
			break;
		case 'w':
			basep = &wflag;
			break;
		case 'l':
			basep = &lflag;
			break;
		default:
			basep = NULL;
		}
		if (basep && *basep & ofmt[i].f_base)
			for (j = 0; j < sizeof align; j++)
				if (ofmt[i].f_align[j] > align[j])
					align[j] = ofmt[i].f_align[j];
	}
}

static void
prnt(FILE *fp, long long of)
{
	wint_t	wc;
	char	b, *mb;
	int	c, lastc = '\n', n;
	long long	rd = 0;

	while ((nflag == 0 || rd < nflag)) {
		if (mb_cur_max > 1) {
			if ((mb = wcget(fp, &wc, &n)) == NULL)
				break;
		} else {
			if ((c = getc(fp)) == EOF)
				break;
			b = wc = c;
			mb = &b;
			n = 1;
		}
		if (lastc == '\n') {
			prna(of);
			putchar('\t');
		}
		of += n, rd += n;
		if (n == 1) {
			c = *mb&0377;
			lastc = c;
			if (wc != WEOF && isprint(c) && c != '\\' &&
					c != '^' && c != '~')
				putchar(c);
			else
				prnc(c);
			if (lastc == '\n')
				putchar('\n');
		} else {
			lastc = c = EOF;
			if (wc != WEOF && iswprint(wc))
				while (n--) {
					putchar(*mb&0377);
					mb++;
				}
			else
				while (n--) {
					prnc(*mb&0377);
					mb++;
				}
		}
	}
	if (lastc != '\n')
		putchar('\n');
	prna(of);
	putchar('\n');
}

static void
prnc(int c)
{
	if (c == 0177 || c == 0377) {
		printf("\\%o", c);
		return;
	}
	if (c & 0200) {
		putchar('~');
		c &= 0177;
	}
	if (c < 040) {
		putchar('^');
		c |= 0100;
	}
	if (c == '\\' || c == '~' || c == '^')
		putchar('\\');
	putchar(c);
}

static char *
wcget(FILE *fp, wint_t *wc, int *len)
{
	static char	mbuf[MB_LEN_MAX+1];
	static char	*mcur, *mend;
	static int	incompl;
	size_t	rest;
	int	c, i, n;

	i = 0;
	rest = mend - mcur;
	if (rest && mcur > mbuf) {
		do
			mbuf[i] = mcur[i];
		while (i++, --rest);
	} else if (incompl) {
		incompl = 0;
		*wc = WEOF;
		mend = mcur = NULL;
		return NULL;
	}
	if (i == 0) {
		c = getc(fp);
		if (c == EOF) {
			*wc = WEOF;
			mend = mcur = NULL;
			return NULL;
		}
		mbuf[i++] = c;
	}
	if (mbuf[0] & 0200) {
		while (mbuf[i-1] != '\n' && i < mb_cur_max &&
				incompl == 0) {
			c = getc(fp);
			if (c != EOF)
				mbuf[i++] = c;
			else
				incompl = 1;
		}
		n = mbtowi(wc, mbuf, i);
		if (n < 0) {
			*len = 1;
			*wc = WEOF;
		} else if (n == 0) {
			*len = 1;
			*wc = '\0';
		} else
			*len = n;
	} else {
		*wc = mbuf[0];
		*len = n = 1;
	}
	mcur = &mbuf[*len];
	mend = &mcur[i - *len];
	return mbuf;
}
